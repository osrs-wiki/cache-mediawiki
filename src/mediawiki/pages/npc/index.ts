export { default as npcPageBuilder } from "./npc";
