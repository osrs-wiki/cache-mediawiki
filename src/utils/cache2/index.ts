export * from "./LazyPromise";
export * from "./types";
