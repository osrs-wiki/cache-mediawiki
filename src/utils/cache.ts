import { DiskCacheProvider, FlatCacheProvider } from "@abextm/cache2";
import { readFile } from "fs/promises";

export type CacheSource = "github" | "local";
export type CacheFileType = "flat" | "disk";

export const getCacheProviderLocal = async (
  version: string,
  type: CacheFileType
) => {
  if (type === "disk") {
    return getCacheProviderDiskLocal(version);
  } else if (type === "flat") {
    return getCacheProviderFlatLocal(version);
  }
};

export const getCacheProviderDiskLocal = async (version: string) => {
  return new DiskCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/diskcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderFlatLocal = async (version: string) => {
  return new FlatCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/flatcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderGithub = async (version = "master") => {
  return new FlatCacheProvider({
    async getFile(name) {
      const req = await fetch(
        `https://raw.githubusercontent.com/abextm/osrs-cache/${version}/${name}`
      );
      if (!req.ok) {
        return;
      }
      const ab = await req.arrayBuffer();
      return new Uint8Array(ab);
    },
  });
};
