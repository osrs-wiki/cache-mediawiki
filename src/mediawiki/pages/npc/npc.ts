import {
  MediaWikiBreak,
  MediaWikiBuilder,
  MediaWikiFile,
  MediaWikiHeader,
  MediaWikiTemplate,
  MediaWikiText,
} from "@osrs-wiki/mediawiki-builder";

import InfoboxMonsterTemplate from "@/mediawiki/templates/InfoboxMonster";
import InfoboxNpcTemplate from "@/mediawiki/templates/InfoboxNpc";
import { NPC } from "@/utils/cache2";

export const npcPageBuilder = (npc: NPC) => {
  const infoboxNpc =
    npc.combatLevel > 0 ? InfoboxMonsterTemplate(npc) : InfoboxNpcTemplate(npc);

  const builder = new MediaWikiBuilder();
  builder.addContents([
    new MediaWikiTemplate("New Content"),
    infoboxNpc.build(),
  ]);

  // Only add chathead image if NPC has chathead models
  if (npc.chatheadModels.length > 0) {
    builder.addContent(
      new MediaWikiFile(`${npc.name} chathead.png`, {
        horizontalAlignment: "left",
      })
    );
  }

  builder.addContents([
    new MediaWikiBreak(),
    new MediaWikiText(npc.name, { bold: true }),
  ]);

  if (npc.actions.includes("Talk-to")) {
    const transcriptTemplate = new MediaWikiTemplate("Hastranscript");
    transcriptTemplate.add("", "npc");
    builder.addContents([
      new MediaWikiBreak(),
      new MediaWikiBreak(),
      new MediaWikiHeader("Dialogue", 2),
      new MediaWikiBreak(),
      transcriptTemplate,
    ]);
  }

  return builder;
};

export default npcPageBuilder;
