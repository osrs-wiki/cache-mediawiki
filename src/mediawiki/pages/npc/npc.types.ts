import { NPC } from "@/utils/cache2";

export type NpcGroup = {
  type: "npc" | "monster";
  npcs: NPC[];
  label: string;
};
