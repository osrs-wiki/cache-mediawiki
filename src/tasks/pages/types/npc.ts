import { writePageToFile } from "../pages.utils";
import { npcPageBuilder } from "../../../mediawiki/pages/npc";
import { CacheProvider, NPC } from "../../../utils/cache2";

export const writeNpcPageFromCache = async (
  cache: Promise<CacheProvider>,
  id: number
) => {
  try {
    const npc = await NPC.load(cache, id);

    if (npc) {
      writeNpcPage(npc);
    }
  } catch (e) {
    console.error(`Error generating page for NPC ${id}: `, e);
  }
};

export const writeNpcPage = async (npc: NPC) => {
  const builder = npcPageBuilder(npc);
  writePageToFile(builder, "npc", npc.name, npc.id.toString());
};
