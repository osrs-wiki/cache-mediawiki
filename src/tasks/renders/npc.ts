import { existsSync } from "fs";
import { copyFile, mkdir } from "fs/promises";

import Context from "../../context";
import { NPC } from "../../utils/cache2";
import { formatFileName } from "../../utils/files";

export const renderNpcs = async (npc: NPC) => {
  if (npc.name.toLocaleLowerCase() === "null") {
    return;
  }
  try {
    if (
      Context.renders &&
      existsSync(`./data/${Context.renders}/npc/${npc.id}.png`)
    ) {
      await mkdir(`./out/${Context.renders}/npc`, { recursive: true });
      await copyFile(
        `./data/${Context.renders}/npc/${npc.id}.png`,
        formatFileName(`./out/${Context.renders}/npc/${npc.name}.png`)
      );
    }
  } catch (e) {
    console.error(e);
  }
};
