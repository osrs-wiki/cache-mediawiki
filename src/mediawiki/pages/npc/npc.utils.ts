import { NpcGroup } from "./npc.types";

import { NPC } from "@/utils/cache2";
import { stripHtmlTags } from "@/utils/string";

export function groupNpcsByType(npcs: NPC[]): NpcGroup[] {
  const monsters = npcs.filter((npc) => npc.combatLevel > 0);
  const nonCombat = npcs.filter((npc) => npc.combatLevel === 0);

  const groups: NpcGroup[] = [];

  if (nonCombat.length > 0) {
    groups.push({
      type: "npc",
      npcs: nonCombat,
      label: generateNpcGroupLabel(nonCombat),
    });
  }

  if (monsters.length > 0) {
    groups.push({
      type: "monster",
      npcs: monsters,
      label: generateMonsterGroupLabel(monsters),
    });
  }

  return groups;
}

function generateNpcGroupLabel(npcs: NPC[]): string {
  // Logic to generate appropriate labels like "Royal servant", "Cultist"
  // Could be based on NPC names, contexts, or other attributes
  const firstNpc = npcs[0];
  if (firstNpc.name && firstNpc.name.toLowerCase() !== "null") {
    return stripHtmlTags(firstNpc.name);
  }
  return "NPC";
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function generateMonsterGroupLabel(npcs: NPC[]): string {
  // Generate labels like "In combat", "Combat form"
  return "In combat";
}
