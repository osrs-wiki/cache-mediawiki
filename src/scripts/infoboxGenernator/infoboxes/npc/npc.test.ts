import fsPromises from "fs/promises";

import npcInfoboxGenerator, {
  buildMonsterInfobox,
  buildNpcInfobox,
} from "./npc";
import * as builders from "./npc";
import { NPC, NPCID } from "../../../../utils/cache2";

describe("NPC Infobox", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe("npcInfoboxGenerator", () => {
    it("should generate monster infobox for an npc with a combat level", async () => {
      jest.spyOn(NPC, "load").mockReturnValue({ id: 1, combatLevel: 1 });
      const buildMonsterInfoboxFn = jest.fn();
      jest
        .spyOn(builders, "buildMonsterInfobox")
        .mockImplementationOnce(buildMonsterInfoboxFn);
      await npcInfoboxGenerator({} as any, 1);
      expect(buildMonsterInfoboxFn).toHaveBeenCalled();
    });

    it("should generate npc infobox for an npc without a combat level", async () => {
      jest.spyOn(NPC, "load").mockReturnValue({
        id: 1,
        combatLevel: 0,
      });
      const buildNpcInfoboxFn = jest.spyOn(builders, "buildNpcInfobox");
      await npcInfoboxGenerator({} as any, 1);
      expect(buildNpcInfoboxFn).toHaveBeenCalled();
    });
  });
  describe("buildNpcInfobox", () => {
    it("should build npc infobox", async () => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore Do not require all fields
      const builder = await buildNpcInfobox({
        name: "name",
        combatLevel: 1,
        actions: ["action1", "action2"],
        id: 1 as NPCID,
      });
      expect(builder?.build()).toMatchSnapshot();
    });
  });

  describe("buildMonsterInfobox", () => {
    it("should build monster infobox", async () => {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore Do not require all fields
      const builder = await buildMonsterInfobox({
        name: "name",
        combatLevel: 100,
        actions: ["action1", "action2"],
        id: 1 as NPCID,
        size: 2,
        hitpoints: 100,
        attack: 200,
        defence: 300,
        magic: 400,
        ranged: 500,
      });
      expect(builder?.build()).toMatchSnapshot();
    });
  });
});
