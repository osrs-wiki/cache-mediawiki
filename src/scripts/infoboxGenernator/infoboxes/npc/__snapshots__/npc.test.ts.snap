// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NPC Infobox buildMonsterInfobox should build monster infobox 1`] = `
"{{New Content}}
{{Infobox Monster
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|combat = 100
|size = 2
|hitpoints = 100
|att = 200
|def = 300
|mage = 400
|range = 500
|examine = 
|id = beta1
}}
'''name'''"
`;

exports[`NPC Infobox buildNpcInfobox should build npc infobox 1`] = `
"{{New Content}}
{{Infobox NPC
|name = name
|image = [[File:name.png|120px]]
|members = Yes
|level = 1
|race = [[Human]]
|location = 
|gender = Male
|options = action1, action2
|map = No
|examine = 
|id = 1
}}
[[File:name chathead.png|left]]
'''name'''"
`;
